compiler/rustc_ty_utils/src/ty.rs

```diff
@@ -35,31 +42,44 @@ fn sized_constraint_for_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Option<Ty<'
         | ty::Never
         | ty::Dynamic(_, _, ty::DynStar) => None,
 
-        // these are never sized
-        ty::Str | ty::Slice(..) | ty::Dynamic(_, _, ty::Dyn) | ty::Foreign(..) => Some(ty),
-
-        ty::Pat(ty, _) => sized_constraint_for_ty(tcx, *ty),
-
-        ty::Tuple(tys) => tys.last().and_then(|&ty| sized_constraint_for_ty(tcx, ty)),
-
-        // recursive case
-        ty::Adt(adt, args) => adt.sized_constraint(tcx).and_then(|intermediate| {
-            let ty = intermediate.instantiate(tcx, args);
-            sized_constraint_for_ty(tcx, ty)
-        }),
+        ty::Str | ty::Slice(..) | ty::Dynamic(_, _, ty::Dyn) => match sizedness {
+            // Never `Sized`
+            SizedTraitKind::Sized => Some(ty),
+            // Always `MetaSized` and `PointeeSized`
+            SizedTraitKind::MetaSized | SizedTraitKind::PointeeSized => None,
+        },
 
-        // these can be sized or unsized.
+        // Maybe `Sized`, `MetaSized` or `PointeeSized`
         ty::Param(..) | ty::Alias(..) | ty::Error(_) => Some(ty),
 
         // We cannot instantiate the binder, so just return the *original* type back,
         // but only if the inner type has a sized constraint. Thus we skip the binder,
         // but don't actually use the result from `sized_constraint_for_ty`.
         ty::UnsafeBinder(inner_ty) => {
-            sized_constraint_for_ty(tcx, inner_ty.skip_binder()).map(|_| ty)
+            sizedness_constraint_for_ty(tcx, sizedness, inner_ty.skip_binder()).map(|_| ty)
+        }
+
+        ty::Foreign(..) => match sizedness {
+            // Never `MetaSized` or `Sized`
+            SizedTraitKind::Sized | SizedTraitKind::MetaSized => Some(ty),
+            // Always `PointeeSized`
+            SizedTraitKind::PointeeSized => None,
+        },
```

Comment:
it's kind of unfortunate that we need to have the same match in multiple places, but it also seems kinda unavoidable

Reply:
Yeah, between the new solver, old solver and `is_trivially_sized`, lots of similar functions, but hard to deduplicate.